### PR TITLE
Added gpu limit for openshift manager

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
This ensures OpenShift manager can schedule pods with GPU requests. This is specific to OpenShift 3.6, 3.7. This has not been tested.

cc @danmcp

### CODE REVIEW

File: pman/openshiftmgr.py [hunk 1]
```diff
@@ -71,6 +71,10 @@ def schedule(self, image, command, name, number_of_workers,
                                     {
                                         "name": "MEMORY_LIMIT",
                                         "value": memory_limit
+                                    },
+                                    {
+                                        "name": "GPU_LIMIT",
+                                        "value": gpu_limit
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: Not sure if this is needed. But added it anyways.

**Comment by danmcp** [comment]
@ravisantoshgudimetla LGTM Let me know if you think this is ready to merge.

File: pman/openshiftmgr.py (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pman/openshiftmgr.py]: Not tested yet. Planning to test it in kumo clusterl

File: pman/openshiftmgr.py [hunk 2]
```diff
@@ -98,6 +102,18 @@ def schedule(self, image, command, name, number_of_workers,
...
+            # The assumption is containers[0] is always image plugin pod node as the publish container is
+            # appended later.
+            # These are all specific to 3.6,3.7. We need to update this once we upgrade to 3.8+, we need to use device
+            # plugin model. Ref: https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/
+            d_job['spec']['template']['spec']['containers'][0]['resources']['limits']['alpha.kubernetes.op/nvidia-gpu'] = gpu_limit
+            d_job['spec']['template']['spec']['containers'][0]['volumeMounts'].append({'mountPath': '/usr/local/nvidia/lib64', 'name': "nvidia-libraries"})
+            d_job['spec']['template']['spec']['volumes'] = [({'name': 'nvidia-libraries', 'hostPath': {'path': '/usr/lib/nvidia-375'}})]
+
+            # TODO: @ravig As of now hardcoding the label key, value here. Key is accelerator and value is gpu-node.
+            # The same labels should be used on node. Remove this hardcoding.
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: Testing it in Kumo cluster.

**Comment by ravisantoshgudimetla** [comment]
@danmcp Made changes for 3.9. It needs testing.

File: pman/openshiftmgr.py [hunk 3]
```diff
@@ -71,7 +71,7 @@ def schedule(self, image, command, name, number_of_workers,
                                     {
                                         "name": "MEMORY_LIMIT",
                                         "value": memory_limit
-                                    }
```
Comment (danmcp) [inline pman/openshiftmgr.py]: why?

File: pman/openshiftmgr.py [hunk 4]
```diff
@@ -98,6 +98,13 @@ def schedule(self, image, command, name, number_of_workers,
                 }
             }
         }
+        if int(gpu_limit) > 0:  # Typecasting before a check
+            # The assumption is containers[0] is always image plugin pod as the publish container is appended later.
+            # These is specific to 3.9+  Ref: https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/
+            d_job['spec']['template']['spec']['containers'][0]['resources']['limits']['nvidia.com/gpu'] = int(gpu_limit)
```
Comment (danmcp) [inline pman/openshiftmgr.py]: How do we take into account clusters that don't have gpus?

File: pman/openshiftmgr.py (hunk 3 above)
Comment (ravisantoshgudimetla) [reply pman/openshiftmgr.py]: Comma addition caused this.

File: pman/openshiftmgr.py (hunk 4 above)
Comment (ravisantoshgudimetla) [reply pman/openshiftmgr.py]: Sorry, did not get the question. If you are asking what will happen if cluster doesn't have gpu enabled nodes, we should not pass gpu_count parameter for the plugin.  I believe if it is passed, then the scope of handling it is beyond openshift mgr.

Comment (danmcp) [reply pman/openshiftmgr.py]: Are you saying the value must be set in CUBE to 0 if the cluster doesn't support GPUs?  If so I think that's acceptable.

Comment (ravisantoshgudimetla) [reply pman/openshiftmgr.py]: That is correct. This piece of code won't have knowledge of what administrator of CUBE configured it to.

**Comment by ravisantoshgudimetla** [comment]
@danmcp Just tested this code successfully on 3.9 cluster. This was able to add the required node selector(accelerator=gpu). This is good to go. PTAL

File: pman/openshiftmgr.py (hunk 3 above)
Comment (danmcp) [reply pman/openshiftmgr.py]: It doesn't seem like a good addition though.

Comment (ravisantoshgudimetla) [reply pman/openshiftmgr.py]: Ack. Done.

**Pull request merged**